### ISSUE TITLE
azurerm_virtual_machine: Support for Spot VM 

### ISSUE THREAD

**Issue by CoRfr** [terraform-providers/terraform-provider-azurerm#5068]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Azure Spot VMs are not in preview as of Dec 3rd ( https://azure.microsoft.com/en-ca/updates/azure-spot-vms-now-in-preview/ ), and the Azure VMSS low-priority VMs, in preview, are deprecated and will be retired on January 1st. This is a very short period of time and it unsure if the existing API to create low-priority VMs will be gracefully transitioned to Spot VMs, or if it will just stop working on Jan 1st. The 'retired' in the Azure documentation makes me think that's option 2. 

#### VMSS

Regarding VMSS, what I can tell from the doc is:
- A new Priority has appeared `Spot` (used to be just `Normal` and `Low`): https://docs.microsoft.com/en-us/rest/api/compute/virtualmachinescalesets/createorupdate#virtualmachineprioritytypes
 - A new BillingProfile can be configured to set a `maxPrice`: https://docs.microsoft.com/en-us/rest/api/compute/virtualmachinescalesets/createorupdate#billingprofile
As of now it seems that variable pricing is not enabled yet.
Also:
> You can set the maxPrice to -1 to indicate that the Azure Spot VM/VMSS should not be evicted for price reasons. Also, the default max price is -1 if it is not provided by you.

#### VM

For VMs, a couple of things are added:
- priority: similar to VMSS, but while that parameter already exists in `azurerm_virtual_machine_scale_set` since low-priority VMSS are already supported, it would be new for `azurerm_virtual_machine`
- billingProfile: seems similar to VMSS
- evictionPolicy: either `Deallocate` or `Delete`, valid for Spot VMs: https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/createorupdate#virtualmachineevictionpolicytypes

### New or Affected Resource(s)

* azurerm_virtual_machine
* azurerm_virtual_machine_scale_set

### Potential Terraform Configuration

```hcl
resource "azurerm_virtual_machine" "node" {
  name                          = "${var.vm_name}"
  location                      = "${var.location}"
  resource_group_name           = "${var.resource_group}"
  vm_size                       = "${var.node_vm_size}"
  network_interface_ids         = ["${azurerm_network_interface.node_nic.*.id[count.index]}"]
  delete_os_disk_on_termination = true

  # New
  priority            = "Spot"
  eviction_policy     = "Deallocate"
  billing_profile {
    max_price = -1
  }

  ...
}

```

### References

* https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/createorupdate
* https://azure.microsoft.com/en-us/pricing/spot/
* https://www.terraform.io/docs/providers/azurerm/r/virtual_machine.html

**Comment by tombuildsstuff** [comment]
hey @CoRfr 

Thanks for opening this issue :)

We're introducing replacements for the `azurerm_virtual_machine` and `azurerm_virtual_machine_scale_set` resources in version 2.0 of the Azure Provider (more details are available in #2807) - as such support for these is being worked on at this time. Since these resources are being superseded - rather than adding this to the existing resources we're instead ensuring that the replacements support this functionality from the start.

At this time we've finished support for the Virtual Machine Scale Set resources (although they're feature toggled off at this time) - and support for the new Virtual Machine resources is being worked on - which in the case of the VMSS resource supports most of the functionality mentioned above. Based on the information above it sounds like there could be some minor modifications required - but we'll ensure that the new resources include these updated fields when we release them in an opt-in Beta in the not-too-distant future 👍

Thanks!

**Comment by CoRfr** [comment]
Given the retirement of the VMSS low-priority feature very soon (Jan. 1st), do you have any idea of when the opt-in Beta would start?

I'm still not sure how the existing API will behave after Jan. 1st, but I'm really concerned it's going to break and that the azurerm provider might not be ready to use the new API.

**Comment by jomeier** [comment]
Hi Terraform Azure Provider team,

because Azure Spot VMs are mission critical for us too we are very interested in Version 2.0 of the Azure Provider (also because of the new timeout feature).

The milestone list for 2.0 indicates that it's rather far away.

Could you give us a hint when a beta or the release might be available?

Thanks a lot for all your effort and greetings,

Josef

**Comment by tombuildsstuff** [comment]
@jomeier from our side the last major feature of 2.0 left to finish is the new VM Resources, which I'm working on at the moment - as such whilst we don't have [an update to the Date on the Github Milestone at this time](https://github.com/terraform-providers/terraform-provider-azurerm/milestone/31), it's worth noting the "percentage complete" shown by Github is kind of misleading due to the number of issues (~half of them) which are tagged as "virtual machine" which'll be solved by this last (major) feature.

Once the new VM resources have been completed we'll publish an update in #2807 - as such I'd recommend subscribing to that issue for updates - but we're intending to launch an opt-in Beta for this functionality once this is finished in the not-too-distant future (and subsequently 2.0 following that).

Thanks!

**Comment by jomeier** [comment]
@tombuildsstuff 
Thanks a lot for the clarification. This helps a lot.

Josef

**Comment by CoRfr** [comment]
I am very concerned that the Azure VMSS Low-priority API will stop working on February 3rd, given that it's scheduled for retirement at that date (see https://azure.microsoft.com/en-ca/updates/azure-spot-vm-meter-changes/ ). There no clear info about what's going to happen on that date, so we're left in the dark.

It doesn't seem like the 2.0 is going to be available by then, and if it is it means that we need to refactor our whole infra around it, and likely face new bugs in prod.
 
Our effort to address the issue by providing Spot VMs on 1.x branch has been rejected ( #5272 ).

I've opened support case 120011824000482 to get some more info around this retirement.

**Comment by tombuildsstuff** [comment]
hey @CoRfr @jomeier

As mentioned above over the past few months we’ve been working on the functionality coming in version 2.0 of the Azure Provider (outlined in #2807). We've just released version 1.43 of the Azure Provider which allows you to opt-in to the Beta of these upcoming features, including the new Virtual Machine and Virtual Machine Scale Set resources.

More details on [how to opt-into the Beta can be found in the Beta guide](https://registry.terraform.io/providers/hashicorp/azurerm/1.43.0/docs/guides/2.0-beta) - however please note that this is only supported in Version 1.43 of the Azure Provider. You can upgrade to this version by updating your Provider block like so:

```
provider "azurerm" {
  version = "=1.43.0"
}
```

and then running `terraform init -upgrade` which will download this version of the Azure Provider.

Once you've [opted into the Beta](https://registry.terraform.io/providers/hashicorp/azurerm/1.43.0/docs/guides/2.0-beta) you can use the new Virtual Machine & Virtual Machine Scale Set resources which support the updated Spot functionality. The documentation for each resource can be found below:

* [`azurerm_linux_virtual_machine`](https://www.terraform.io/docs/providers/azurerm/r/linux_virtual_machine.html)
* [`azurerm_linux_virtual_machine_scale_set`](https://www.terraform.io/docs/providers/azurerm/r/linux_virtual_machine_scale_set.html)
* [`azurerm_virtual_machine_scale_set_extension`](https://www.terraform.io/docs/providers/azurerm/r/virtual_machine_scale_set_extension.html)
* [`azurerm_windows_virtual_machine`](https://www.terraform.io/docs/providers/azurerm/r/windows_virtual_machine.html)
* [`azurerm_windows_virtual_machine_scale_set`](https://www.terraform.io/docs/providers/azurerm/r/windows_virtual_machine_scale_set.html)

Unfortunately we have no plans to backport this to the existing `azurerm_virtual_machine` and `azurerm_virtual_machine_scale_set` resources which are being superseded in 2.0 (and as such are in a feature-frozen state) - however the new resources do support this functionality.

---

The new Virtual Machine & Virtual Machine Scale Set resources will be going GA with Version 2.0 of the Azure Provider in the coming weeks - if you've tried the Beta and have feedback [please open a Github Issue using the special `Beta Feedback` category](https://github.com/terraform-providers/terraform-provider-azurerm/issues/new?template=Beta_Feedback.md) and we'll take a look.

Thanks!
